### ISSUE TITLE
[fs2.io.net] `read` after a cancelled `read` causes `java.nio.channels.ReadPendingException`

### ISSUE THREAD

**Issue by kory33** [typelevel/fs2#2709]
## Issue description
Given a TCP client `client: Socket[cats.effect.IO]`, a `client.read` immediately after a cancelled `client.read` (sometimes) causes `java.nio.channels.ReadPendingException`.

## Reproduction
On Windows 10, OpenJDK 17, prepare the following files:

<details>
  <summary>build.sbt</summary>

```sbt
scalaVersion := "3.1.0"
libraryDependencies ++= Seq(
  "co.fs2" %% "fs2-io" % "3.2.2"
)
```

</details>

<details>
  <summary>src/main/scala/Test.scala (binds port 31629, change portNumber if needed)</summary>

```Scala
object Test {
  @main def main: Unit = {
    import cats.Monad
    import cats.effect.{IO, Resource, Spawn}
    import com.comcast.ip4s.SocketAddress
    import fs2.io.net.Network

    import cats.implicits.given
    import cats.effect.unsafe.implicits.global

    val portNumber = 31629
    val serverAddress = SocketAddress.fromString(s"localhost:$portNumber").get
    val clientResource: Resource[IO, fs2.io.net.Socket[IO]] =
      for {
        // we prepare a tcp server which returns 0 for every read from a client
        _ <- Spawn[IO].background {
          Network[IO]
            .server(address = Some(serverAddress.host), port = Some(serverAddress.port))
            .evalMap(socket =>
              Monad[IO].whileM_(socket.isOpen)(socket.write(fs2.Chunk[Byte](0)))
            )
            .compile
            .drain
        }
        socket <- Network[IO].client(serverAddress)
      } yield socket

    val program: IO[Unit] =
      clientResource.use { client =>
        Monad[IO].foreverM {
          client.read(1024).start.flatMap { fiber =>
            fiber.cancel >> client.read(1024) >> fiber.join
          }
        }
      }.void

    program.unsafeRunSync()
  }
}
```

</details>

And `runMain main` in `sbt` console.

<details>
  <summary>example log</summary>

```
[IJ]runMain main
[info] running main
[error] (run-main-1) java.nio.channels.ReadPendingException
[error] java.nio.channels.ReadPendingException
[error]         at java.base/sun.nio.ch.AsynchronousSocketChannelImpl.read(AsynchronousSocketChannelImpl.java:251)
[error]         at java.base/sun.nio.ch.AsynchronousSocketChannelImpl.read(AsynchronousSocketChannelImpl.java:297)
[error]         at java.base/java.nio.channels.AsynchronousSocketChannel.read(AsynchronousSocketChannel.java:425)
[error]         at fs2.io.net.SocketCompanionPlatform$AsyncSocket.readChunk$$anonfun$1(SocketPlatform.scala:120)
[error]         at async_ @ fs2.io.net.SocketCompanionPlatform$AsyncSocket.readChunk(SocketPlatform.scala:121)
[error]         at flatMap @ fs2.io.net.SocketCompanionPlatform$BufferedReads.read$$anonfun$1(SocketPlatform.scala:87)
[error]         at delay @ fs2.io.net.SocketCompanionPlatform$BufferedReads.withReadBuffer$$anonfun$1(SocketPlatform.scala:60)
[error]         at flatMap @ fs2.Compiler$Target.flatMap(Compiler.scala:162)
[error]         at modify @ fs2.internal.Scope.close(Scope.scala:262)
[error]         at flatMap @ fs2.Compiler$Target.flatMap(Compiler.scala:162)
[error] stack trace is suppressed; run 'last Compile / bgRunMain' for the full output
[error] Nonzero exit code: 1
[error] (Compile / runMain) Nonzero exit code: 1
[error] Total time: 2 s, completed 2021/11/14 23:29:13
[IJ]
```

</details>

## Notes
 - For some reason I could not get this work on [scastie](https://scastie.scala-lang.org/v92TgbhcQk6UGVfuBfdu4Q) (it times out...) A full project example can be found [here](https://github.com/kory33/fs2-bug-reports/tree/28bfe4534eb499084f2f27dd8d207f61f0b6771e).
 - The issue persists with `readN` as well, but it takes some more time for exception to come out with larger `max`
 - My guess is that [`AsyncSocket.readChunk`](https://github.com/typelevel/fs2/blob/a7f0abc0ac1df68ba7ac78f9386f1c1acd52c69d/io/jvm/src/main/scala/fs2/io/net/SocketPlatform.scala#L116-L122) should hold `readSemaphore` permit until `IntCallbackHandler` has been invoked, and not until a read operation is complete.

**Comment by mpilquist** [comment]
Hm, `readChunk` is only ever called while the `readSempahore` is held -- inside a call to `withReadBuffer`:
```scala
    private def withReadBuffer[A](size: Int)(f: ByteBuffer => F[A]): F[A] =
      readSemaphore.permit.use { _ =>
        F.delay {
          if (readBuffer.capacity() < size)
            readBuffer = ByteBuffer.allocateDirect(size)
          else
            (readBuffer: Buffer).limit(size)
          f(readBuffer)
        }.flatten
      }
```

Because of that `.flatten` on the second to last line there, we hold the read semaphore until the flattened task completes -- i.e., until an `A` value is produced -- which is necessarily after `IntCallbackHandler` completes. I must be missing something though...

**Comment by mpilquist** [comment]
Oh I see -- when a fiber that's waiting on a socket read is cancelled, the call to `ch.read` is left running. This is because it's wrapped in `F.async_` instead of `F.async`. So the `ch.read` remains pending and the semaphore is released soon afterwards.

To fix this, we'd need some way to interrupt the pending `ch.read` operation. I'm not sure that's possible with the NIO APIs.

**Comment by mpilquist** [comment]
@armanbilge had another idea - we could not release the read semaphore if the task is cancelled. This turns any request to read after cancelation in to hangs instead of exceptions.

**Comment by armanbilge** [comment]
More specifically, I think the read semaphore should be released by the asynchronous read (which should eventually return IIUC, seeing as we cannot cancel it). I think this is essentially what @kory33 suggested:

> My guess is that [`AsyncSocket.readChunk`](https://github.com/typelevel/fs2/blob/a7f0abc0ac1df68ba7ac78f9386f1c1acd52c69d/io/jvm/src/main/scala/fs2/io/net/SocketPlatform.scala#L116-L122) should hold `readSemaphore` permit until `IntCallbackHandler` has been invoked, and not until a read operation is complete.

**Comment by mpilquist** [comment]
The completion handler cannot release the semaphore as we are using a shared buffer.

**Comment by mpilquist** [comment]
This is the change I was suggesting, though I don't like the result of ending up with a hung task.

```scala
    private def withReadBuffer[A](size: Int)(f: ByteBuffer => F[A]): F[A] =
      Resource.makeCase(readSemaphore.acquire) {
        case (_, Resource.ExitCase.Canceled) => F.unit
        case _ => readSemaphore.release
      }.use { _ =>
        F.delay {
          if (readBuffer.capacity() < size)
            readBuffer = ByteBuffer.allocateDirect(size)
          else
            (readBuffer: Buffer).limit(size)
          f(readBuffer)
        }.flatten
      }
```

**Comment by armanbilge** [comment]
Hmm, I didn't realize that. Not sure how much I like this anymore either.

**Comment by armanbilge** [comment]
@mpilquist sorry, can you explain more why this is an issue?

> The completion handler cannot release the semaphore as we are using a shared buffer.

Isn't the shared buffer what the semaphore is guarding access to?

**Comment by mpilquist** [comment]
Right, but we're not done with the buffer when in the completion handler. While holding that same semaphore, we dump the buffer to a chunk.

**Comment by armanbilge** [comment]
I see, thanks. Can we do a two-step release? Where both the handler needs to be invoked, and the current release we already have?

**Comment by mpilquist** [comment]
Maybe? Not sure what that would accomplish really. The more I think about this, the more I think failing with an exception is the right behavior assuming there's no way to get NIO to do what we want.

**Comment by armanbilge** [comment]
It's because there's two things we need to guard:
1. access to the shared buffer
2. whether there is a pending read

In order to read, you need to acquire permission for both. But, they are released at separate times, (1) after dumping to a chunk and (2) after the callback completes.

**Comment by armanbilge** [comment]
> The more I think about this, the more I think failing with an exception is the right behavior assuming there's no way to get NIO to do what we want.

I'm still of the opinion that this is a truly uncancellable operation (where "getting NIO to do what we want" = cancelling it) and should be treated as such. Any assumptions made by the resulting API that it _is_ cancellable were flawed.

**Comment by kory33** [comment]
I thought something like the following works (where `channelReadSemaphore` is distinct from `readSemaphore`)

```Scala
    protected def readChunk(buffer: ByteBuffer): F[Int] =
      // .background.use does not work because it cancels the spawned fiber on cancellation
      F.start(
        channelReadSemaphore.permit.use(_ =>
          F.async_[Int](cb => ch.read(buffer, null, new IntCallbackHandler(cb)))
        )
      ).flatMap(_.join)
        .flatMap(_.embedNever /* the child fiber is never cancelled */ )
```

However, the problem with the above is that access to `buffer` is not protected in the spawned fiber, and another `read` may as well overwrite `buffer` field with a fresh one (or `.limit` the same instance), potentially losing bytes read in the previously spawned fiber.

Given this kind of execution flow I think "carrying over" the result from the unreleased buffer is essential, so maybe we need to modify `withReadBuffer` too?

**Comment by kory33** [comment]
I think making `read` cancelation-safe is in general costly (as in #2710), unless NIO provides a real way to cancel a read. If this is the case I think the best thing to do is either
 - making `read` / `readN` `uncancelable`, or
 - throwing a possible exception on cancelation (current behaviour)

and stating in the doc as such. (I think the former is better, do you have any opinions?)

For a workaround, I wrapped `Socket` `Resource`s to another `Resource` that holds an atomic reference to the buffer with a background process to continuously read from the socket ([code](https://github.com/kory33/s2mc-test/blob/5a444cc099656da2a58adb24d81a731095dfdae9/protocol-impl/src/main/scala/io/github/kory33/s2mctest/impl/generic/net/AtomicSocket.scala)), and this looks alright.

**Comment by mpilquist** [comment]
I think `uncancelable` is the best option. My only concern there is whether we can shutdown a socket that has a pending read. I haven't gotten a chance to test that yet but I think/hope the answer is yes.

**Comment by armanbilge** [comment]
👍 to `uncancelable`, that was my first thought.

**Comment by kory33** [comment]
> My only concern there is whether we can shutdown a socket that has a pending read.

I may be missing something, but shoudn't `uncancelable` eliminate the possibility of pending reads whenever a semaphore's permit goes out of scope?
